File: Scenarios-Python/scenario_1_run_and_wait.py
"""
Scenario 1: Run Task and Wait for Completion ⭐ Most Common

Business Need:
Start a workflow task, wait for it to complete, and then get the results.

Use Cases:
- Need to synchronously wait for task completion
- Need to get the complete output results of the task
- Suitable for single task execution

Usage Steps:
1. Modify API_KEY and WORKFLOW_ID below
2. Modify input_parameters according to your workflow
3. Run script: python Scenarios-Python/scenario_1_run_and_wait.py
"""

import os
import time
import traceback
import json
import requests

# ============ Configuration Area - Please modify the following variables ============
# API Key, get from: https://www.browseract.com/reception/integrations
API_KEY = os.getenv("BROWSERACT_API_KEY", "app-abcdefghijklmn")

# Workflow ID, get from: https://www.browseract.com/reception/workflow-list
WORKFLOW_ID = 1234567890

# Workflow input parameters (modify according to your workflow definition)
INPUT_PARAMETERS = [
    {
        "name": "target_url",
        "value": "https://www.google.com/search?q=iphone17",
    },
    {
        "name": "product_limit",
        "value": "10",
    }
]

# Polling configuration
POLL_INTERVAL = 5  # Check task status every 5 seconds
MAX_WAIT_TIME = 600  # Maximum wait time: 10 minutes (600 seconds)
# ================================================

API_BASE_URL = "https://api.browseract.com/v2/workflow"

def run_task(workflow_id, input_parameters):
    """Start a task"""
    headers = {
        "Authorization": f"Bearer {API_KEY}"
    }
    
    data = {
        "workflow_id": workflow_id,
        "input_parameters": input_parameters,
        "save_browser_data": True,
    }
    
    api_url = f"{API_BASE_URL}/run-task"
    response = requests.post(api_url, json=data, headers=headers)
    
    if response.status_code == 200:
        result = response.json()
        task_id = result["id"]
        print(f"✅ Task started, Task ID: {task_id}")
        if "profileId" in result:
            print(f"   Profile ID: {result['profileId']}")
        return task_id
    else:
        print(f"❌ Failed to start task: {response.json()}")
        return None

def get_task_status(task_id):
    """Get task status"""
    headers = {
        "Authorization": f"Bearer {API_KEY}"
    }
    
    api_url = f"{API_BASE_URL}/get-task-status?task_id={task_id}"
    response = requests.get(api_url, headers=headers)
    
    if response.status_code == 200:
        return response.json().get("status")
    else:
        print(f"⚠️ Failed to get task status: {response.json()}")
        return None

def get_task(task_id):
    """Get detailed task information"""
    headers = {
        "Authorization": f"Bearer {API_KEY}"
    }
    
    api_url = f"{API_BASE_URL}/get-task?task_id={task_id}"
    response = requests.get(api_url, headers=headers)
    
    if response.status_code == 200:
        return response.json()
    else:
        print(f"⚠️ Failed to get task details: {response.json()}")
        return None

def wait_for_task_completion(task_id):
    """Wait for task completion"""
    start_time = time.time()
    
    print(f"\n⏳ Waiting for task completion (max wait time: {MAX_WAIT_TIME} seconds)...")
    
    while True:
        # Check if timeout
        elapsed_time = time.time() - start_time
        if elapsed_time > MAX_WAIT_TIME:
            print(f"\n⏰ Wait timeout (waited {elapsed_time:.0f} seconds)")
            return None
        
        # Get task status
        status = get_task_status(task_id)
        
        if status is None:
            print("⚠️ Unable to get task status, continuing to wait...")
        elif status == "finished":
            print(f"\n✅ Task completed!")
            return "finished"
        elif status == "failed":
            print(f"\n❌ Task execution failed")
            return "failed"
        elif status == "canceled":
            print(f"\n🚫 Task canceled")
            return "canceled"
        else:
            # running, created, paused, etc.
            elapsed = int(elapsed_time)
            print(f"   Status: {status} (waited {elapsed} seconds)", end="\r")
        
        # Wait before checking again
        time.sleep(POLL_INTERVAL)

def main():
    print("=" * 60)
    print("Scenario 1: Run Task and Wait for Completion")
    print("=" * 60)
    
    try:
        # Step 1: Start task
        print("\n📤 Step 1: Starting task...")
        task_id = run_task(WORKFLOW_ID, INPUT_PARAMETERS)
        
        if task_id is None:
            print("❌ Unable to start task, exiting")
            return
        
        # Step 2: Wait for task completion
        print("\n📥 Step 2: Waiting for task completion...")
        final_status = wait_for_task_completion(task_id)
        
        # Step 3: Get task results
        print("\n📊 Step 3: Getting task results...")
        task_info = get_task(task_id)
        
        if task_info:
            print("\n" + "=" * 60)
            print("Task Result (JSON):")
            print("=" * 60)
            # Output complete task result as formatted JSON
            print(json.dumps(task_info, indent=2, ensure_ascii=False))
            print("=" * 60)
        else:
            print("⚠️ Unable to get task details")
            
    except Exception as e:
        error = traceback.format_exc()
        print(f"\n❌ Error occurred:\n{error}")

if __name__ == "__main__":
    main()

File: Scenarios-Java/src/main/java/com/browseract/scenarios/Scenario1RunAndWait.java
package com.browseract.scenarios;

/**
 * Scenario 1: Run Task and Wait for Completion ⭐ Most Common
 * 
 * Business Need:
 * Start a workflow task, wait for it to complete, and then get the results.
 * 
 * Use Cases:
 * - Need to synchronously wait for task completion
 * - Need to get the complete output results of the task
 * - Suitable for single task execution
 * 
 * Usage Steps:
 * 1. Modify API_KEY and WORKFLOW_ID below
 * 2. Modify input_parameters according to your workflow
 * 3. Run: mvn exec:java -Dexec.mainClass="com.browseract.scenarios.Scenario1RunAndWait"
 */

import com.browseract.scenarios.util.HttpUtil;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Scenario1RunAndWait {
    
    // ============ Configuration Area - Please modify the following variables ============
    // API Key, get from: https://www.browseract.com/reception/integrations
    private static final String API_KEY = System.getenv("BROWSERACT_API_KEY") != null 
            ? System.getenv("BROWSERACT_API_KEY") 
            : "app-abcdefghijklmn";
    
    // Workflow ID, get from: https://www.browseract.com/reception/workflow-list
    private static final long WORKFLOW_ID = 1234567890L;
    
    // Polling configuration
    private static final int POLL_INTERVAL = 5;  // Check task status every 5 seconds
    private static final int MAX_WAIT_TIME = 1800;  // Maximum wait time: 30 minutes (1800 seconds)
    // ================================================
    
    private static final String API_BASE_URL = "https://api.browseract.com/v2/workflow";
    private static final ObjectMapper objectMapper = new ObjectMapper();
    
    public static void main(String[] args) {
        System.out.println("============================================================");
        System.out.println("Scenario 1: Run Task and Wait for Completion");
        System.out.println("============================================================");
        
        try {
            // Step 1: Start task
            System.out.println("\n📤 Step 1: Starting task...");
            
            // Workflow input parameters (modify according to your workflow definition)
            List<InputParameter> inputParameters = new ArrayList<>();
            inputParameters.add(new InputParameter("target_url", "https://www.google.com/search?q=iphone17"));
            inputParameters.add(new InputParameter("product_limit", "10"));
            
            String taskId = runTask(WORKFLOW_ID, inputParameters);
            
            if (taskId == null) {
                System.out.println("❌ Unable to start task, exiting");
                return;
            }
            
            // Step 2: Wait for task completion
            System.out.println("\n📥 Step 2: Waiting for task completion...");
            String finalStatus = waitForTaskCompletion(taskId);
            
            // Step 3: Get task results
            System.out.println("\n📊 Step 3: Getting task results...");
            String taskInfoJson = getTask(taskId);
            
            if (taskInfoJson != null) {
                System.out.println("\n============================================================");
                System.out.println("Task Result (JSON):");
                System.out.println("============================================================");
                // Output complete task result as formatted JSON
                JsonNode jsonNode = objectMapper.readTree(taskInfoJson);
                System.out.println(objectMapper.writerWithDefaultPrettyPrinter().writeValueAsString(jsonNode));
                System.out.println("============================================================");
            } else {
                System.out.println("⚠️ Unable to get task details");
            }
            
        } catch (Exception e) {
            System.err.println("\n❌ Error occurred:");
            e.printStackTrace();
        }
    }
    
    private static String runTask(long workflowId, List<InputParameter> inputParameters) {
        Map<String, String> headers = new HashMap<>();
        headers.put("Authorization", "Bearer " + API_KEY);
        
        RunTaskRequest requestBody = new RunTaskRequest();
        requestBody.setWorkflow_id(workflowId);
        requestBody.setInput_parameters(inputParameters);
        requestBody.setSave_browser_data(true);
        
        String apiUrl = API_BASE_URL + "/run-task";
        HttpUtil.HttpResult result = HttpUtil.postJson(apiUrl, requestBody, headers);
        
        if (result.isSuccess()) {
            try {
                JsonNode jsonNode = objectMapper.readTree(result.getText());
                String taskId = jsonNode.get("id").asText();
                System.out.println("✅ Task started, Task ID: " + taskId);
                if (jsonNode.has("profileId")) {
                    System.out.println("   Profile ID: " + jsonNode.get("profileId").asText());
                }
                return taskId;
            } catch (Exception e) {
                System.out.println("❌ Failed to parse response: " + e.getMessage());
                return null;
            }
        } else {
            System.out.println("❌ Failed to start task: " + result.getText());
            return null;
        }
    }
    
    private static String getTaskStatus(String taskId) {
        Map<String, String> headers = new HashMap<>();
        headers.put("Authorization", "Bearer " + API_KEY);
        
        String apiUrl = API_BASE_URL + "/get-task-status?task_id=" + taskId;
        HttpUtil.HttpResult result = HttpUtil.get(apiUrl, headers);
        
        if (result.isSuccess()) {
            try {
                JsonNode jsonNode = objectMapper.readTree(result.getText());
                return jsonNode.get("status").asText();
            } catch (Exception e) {
                return null;
            }
        } else {
            // Network error or API error, will retry in next polling cycle
            return null;
        }
    }
    
    private static String getTask(String taskId) {
        Map<String, String> headers = new HashMap<>();
        headers.put("Authorization", "Bearer " + API_KEY);
        
        String apiUrl = API_BASE_URL + "/get-task?task_id=" + taskId;
        HttpUtil.HttpResult result = HttpUtil.get(apiUrl, headers);
        
        if (result.isSuccess()) {
            return result.getText();
        } else {
            System.out.println("⚠️ Network error while getting task details: " + result.getText());
            return null;
        }
    }
    
    private static String waitForTaskCompletion(String taskId) {
        long startTime = System.currentTimeMillis();
        String previousStatus = null;
        
        System.out.println("\n⏳ Waiting for task completion (max wait time: " + (MAX_WAIT_TIME / 60) + " minutes)...");
        
        while (true) {
            // Check if timeout
            long elapsedTime = (System.currentTimeMillis() - startTime) / 1000;
            if (elapsedTime > MAX_WAIT_TIME) {
                System.out.println("\n⏰ Wait timeout (waited " + elapsedTime + " seconds)");
                return null;
            }
            
            // Get task status
            String status = getTaskStatus(taskId);
            
            if (status == null) {
                // Network error or API error, continue waiting
                System.out.println();
                System.out.print("   Network error, retrying... (waited " + elapsedTime + " seconds)\r");
            } else if ("finished".equals(status)) {
                System.out.println("\n✅ Task completed!");
                return "finished";
            } else if ("failed".equals(status)) {
                System.out.println("\n❌ Task execution failed");
                return "failed";
            } else if ("canceled".equals(status)) {
                System.out.println("\n🚫 Task canceled");
                return "canceled";
            } else {
                // running, created, paused, etc.
                // If status changed, print on new line; otherwise update same line
                if (!status.equals(previousStatus)) {
                    System.out.println();  // New line when status changes
                    System.out.print("   Status: " + status + " (waited " + elapsedTime + " seconds)\r");
                    previousStatus = status;
                } else {
                    // Same status, update same line
                    System.out.print("   Status: " + status + " (waited " + elapsedTime + " seconds)\r");
                }
            }
            
            // Wait before checking again
            try {
                Thread.sleep(POLL_INTERVAL * 1000);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                return null;
            }
        }
    }
    
    // Inner classes for request/response
    public static class RunTaskRequest {
        private long workflow_id;
        private List<InputParameter> input_parameters;
        private boolean save_browser_data;
        
        public long getWorkflow_id() { return workflow_id; }
        public void setWorkflow_id(long workflow_id) { this.workflow_id = workflow_id; }
        
        public List<InputParameter> getInput_parameters() { return input_parameters; }
        public void setInput_parameters(List<InputParameter> input_parameters) { this.input_parameters = input_parameters; }
        
        public boolean isSave_browser_data() { return save_browser_data; }
        public void setSave_browser_data(boolean save_browser_data) { this.save_browser_data = save_browser_data; }
    }
    
    public static class InputParameter {
        private String name;
        private String value;
        
        public InputParameter(String name, String value) {
            this.name = name;
            this.value = value;
        }
        
        public String getName() { return name; }
        public void setName(String name) { this.name = name; }
        
        public String getValue() { return value; }
        public void setValue(String value) { this.value = value; }
    }
}
